### PR TITLE
Change fix to use non-zero exit code if unfixable

### LINKED ISSUE #1340: Return non zero exit code on fix when there are lint violations
We use [lefthook](https://github.com/evilmartians/lefthook) in combination with a pre-commit git hook to fix and commit updated dbt models when committing changes to the repository. This works great with the only downside that `poetry run sqlfluff --fix` will return a 0 exit code when it's done fixing even though there are still unfixable lint violations. 

As an enhancement if would be great if sqlfluff could return a non zero exit code on fix when there are still unfixable linting violations.

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made
Fixes #1340 

It also adds a couple of test cases for recently merged changes (#1382 and #1385) that did not have coverage (we to figure out why Coverage checking has stopped! See #1388):

### Are there any other side effects of this change that we should be aware of?
We now exit `fix` with a non-zero exit code if there are unfixable errors. This is a change in behaviour and may be unexpected by some, but we think it's better to raise these issues.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/cli/commands.py [hunk 1]
```diff
@@ -483,13 +484,14 @@ def fix(force, paths, processes, bench=False, fixed_suffix="", logger=None, **kw
...
-                    click.echo("Fix aborted due to unparseable template variables.")
-                if unfixable_error:
-                    click.echo("Unfixable violations detected.")
             stdout = stdin
 
+        if verbose:
+            if templater_error:
+                click.echo("Fix aborted due to unparseable template variables.")
+            if unfixable_error:
+                click.echo("Unfixable violations detected.")
```
Comment (tunetheweb) [inline src/sqlfluff/cli/commands.py:493]: I thought better to raise this error even if there are fixable errors.

File: src/sqlfluff/cli/commands.py [hunk 2]
```diff
@@ -551,19 +553,31 @@ def fix(force, paths, processes, bench=False, fixed_suffix="", logger=None, **kw
...
-            )
         _completion_message(config)
 
+    if result.num_violations(types=SQLLintError, fixable=False) > 0:
+        click.echo(
+            "  [{} unfixable linting violations found]".format(
+                result.num_violations(types=SQLLintError, fixable=False)
+            )
+        )
+        exit_code = 1
```
Comment (tunetheweb) [inline src/sqlfluff/cli/commands.py:571]: Again I don't like that this is hidden if there are fixable errors so have moved it out of the else statement

File: test/cli/commands_test.py [hunk 3]
```diff
@@ -407,37 +437,60 @@ def test__cli__command_fix_stdin_safety():
...
+            "Fix aborted due to unparseable template variables.",
+        ),  # template error
+        ("create TABLE a.t (a int)", 0, "", ""),  # fixable error
+        ("create table a.t (a int)", 0, "", ""),  # perfection
+        (
+            "select col from a join b using (c)",
+            1,
+            "-v",
+            "Unfixable violations detected.",
+        ),  # unfixable error (using)
```
Comment (tunetheweb) [inline test/cli/commands_test.py:455]: @nolanbconaway I enhanced these tests to get the coverage back up to 100%

File: test/fixtures/rules/std_rule_cases/L016.yml [hunk 4]
```diff
@@ -51,6 +51,22 @@ test_fail_line_too_long_with_comments_4:
...
+    SELECT
+    c1
+    ,--  the "y variable" and uses_small_subject_line to be the "x variable" in terms of the regression line.
+    c2
+  configs:
+    rules:
+      max_line_length: 80
+      L016:
+        ignore_comment_lines: True
+
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L016.yml:69]: @barrywhart we needed to add this test from your recently merged change to get coverage back up to 100%

**Review by tunetheweb — COMMENTED** [review]
Comments below.

@nolanbconaway you were right to leave out of the other PR - the fix itself was easy but the test cases nearly broke my heart 😢 Got there in the end though...

File: src/sqlfluff/cli/commands.py (hunk 1 above)
Comment (nolanbconaway) [reply src/sqlfluff/cli/commands.py:493]: AH yeah good call! That was the real spirit of the issue anyway.

File: test/cli/commands_test.py [hunk 5]
```diff
@@ -20,7 +20,12 @@
 
 
 def invoke_assert_code(
-    ret_code=0, args=None, kwargs=None, cli_input=None, mix_stderr=True
+    ret_code=0,
+    args=None,
+    kwargs=None,
+    cli_input=None,
+    mix_stderr=True,
+    output_contains="",
```
Comment (nolanbconaway) [inline test/cli/commands_test.py:28]: lol i saw it coming

File: test/cli/commands_test.py (hunk 3 above)
Comment (nolanbconaway) [reply test/cli/commands_test.py:455]: Yeah thx! my setup of adding new adhoc strings to test wasnt exactly scalable and this is

**Review by nolanbconaway — APPROVED** [review]
LGTM. thx for paying off my debt @tunetheweb

**Comment by chlirre** [comment]
A `--nofail` option would have been great. See [lint commend docs](https://docs.sqlfluff.com/en/stable/cli.html#cmdoption-sqlfluff-lint-nofail).

I'm attempting to run `fix` as part of a manually triggered Github action to "fix whatever can be fixed" on PRs. The non-zero exit code is, well, not helping 😄

**Pull request merged**